### PR TITLE
!str #16416,#16994 BidiFlow DSL and Flow.join Mat

### PR DESCRIPTION
**Pull request by rkuhn**
- add BidiFlow, with atop and join combinators
- add Flow.join(BidiFlow)
- correct Flow.join’s default materialized value selection to Keep.left
- [x] add factory methods
- [x] add javadsl
- [x] add tests
- [x] add docs

References #16416 #16994

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/scaladsl/BidiFlow.scala [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.scaladsl
+
+import akka.stream.Graph
+import akka.stream.BidiShape
+import akka.stream.impl.StreamLayout.Module
+import akka.stream.FlowShape
+
+final class BidiFlow[-I1, +O1, -I2, +O2, +Mat](private[stream] override val module: Module) extends Graph[BidiShape[I1, O1, I2, O2], Mat] {
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/BidiFlow.scala]: uuu, it's here finally :-)

**Comment by patriknw** [comment]
nice

**Comment by bantonsson** [comment]
Very nice indeed. ASCII art FTW!

**Comment by rkuhn** [comment]
yeah, I was a little bored while de- and re-registering my car ;-)

File: akka-stream/src/main/scala/akka/stream/scaladsl/BidiFlow.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/BidiFlow.scala]: We probably need a factory from a Graph[BidiShape, _]. The same should be added to Flow, Source and Sink (maybe RunnableFlow, too, but that is less useful).

**Comment by drewhk** [comment]
LGTM, after the factory is added.

File: akka-stream/src/main/scala/akka/stream/scaladsl/BidiFlow.scala (hunk 1 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/BidiFlow.scala]: Good point, ISTR adding such a `wrap` method to `Flow`, but it should be on all of them.

**Comment by rkuhn** [comment]
test & docs added, also for Java, please review

File: akka-stream/src/main/scala/akka/stream/scaladsl/JavaConverters.scala [hunk 2]
```diff
@@ -17,6 +17,9 @@ private[akka] object JavaConverters {
   implicit final class AddAsJavaFlow[In, Out, Mat](val flow: scaladsl.Flow[In, Out, Mat]) extends AnyVal {
     def asJava: javadsl.Flow[In, Out, Mat] = new javadsl.Flow(flow)
   }
+  implicit final class AddAsJavaBidiFlow[I1, O1, I2, O2, Mat](val flow: scaladsl.BidiFlow[I1, O1, I2, O2, Mat]) extends AnyVal {
+    def asJava: javadsl.BidiFlow[I1, O1, I2, O2, Mat] = new javadsl.BidiFlow(flow)
+  }
```
Comment (bantonsson) [inline akka-stream/src/main/scala/akka/stream/scaladsl/JavaConverters.scala]: @patriknw has removed the implicit converters here #17019

**Comment by bantonsson** [comment]
LGTM!

The validation failure is just `Sink.head()` and friends loosing their parenthesis.

**Pull request merged**